### PR TITLE
QueryCache: Support queries with totals and extremes

### LINKED ISSUE #48677: Results WITH TOTALS were not cached in Query Cache
> You have to provide the following information whenever possible.

**Describe what's wrong**
When WITH TOTALS and Query Cache are used in queries, the results returned by query cache do not contain the WITH TOTALS results.

**Before the query cache takes effect**
 :) select t1,count() from t5 group by t1 with totals order by t1 desc limit 11

┌─t1─┬─count()─┐
│ aa │      28 │
└────┴─────────┘

Totals:
┌─t1─┬─count()─┐
│    │      28 │
└────┴─────────┘

**After the query cache takes effect**

 :) select t1,count() from t5 group by t1 with totals order by t1 desc limit 11

┌─t1─┬─count()─┐
│ aa │      28 │
└────┴─────────┘

### PR DESCRIPTION
**Pull request by rschu1ze**
Fixes: #48677

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
The query cache can now supports queries with totals and extremes modifier.

### CODE REVIEW

**Comment by rschu1ze** [comment]
Looks like I made pretty much every mistake with pipeline processors one can make but the latest version seems to work. Let me rebase, cleanup and add a test.

**Comment by rschu1ze** [comment]
@nikitat You are the only reviewer I can think of for this PR. Would you like to have a look? Thanks!

The logic is admittedly a bit tricky. In case of questions, please ask.

**Comment by alexey-milovidov** [comment]
@rschu1ze, is it related to the query cache?

https://s3.amazonaws.com/clickhouse-test-reports/48853/4dceee0dc321445ef420b5e37aac69805b40fcc7/upgrade_check__msan_/logical_errors.txt

File: src/Processors/Sources/SourceFromChunks.cpp [hunk 1]
```diff
@@ -3,20 +3,61 @@
...
+
+        if (status != Status::Finished)
+            return status;
+
+        finished_chunks = true;
+    }
+
+    if (getTotalsPort())
+    {
+        /// This logic force-pushes the data into the port instead of checking the port status first (isFinished(), canPush()). Seems to
```
Comment (nickitat) [inline src/Processors/Sources/SourceFromChunks.cpp:40]: it seems easy to fix, let's do it straight away

Comment (rschu1ze) [reply src/Processors/Sources/SourceFromChunks.cpp:40]: I simplified further, that logic is gone now.

**Comment by robot-ch-test-poll4** [comment]
This is an automated comment for commit 0c49cfa743af5aea65be5eaadd796e002485f403 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/48853/0c49cfa743af5aea65be5eaadd796e002485f403/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Review by nickitat — APPROVED** [review]

File: src/Interpreters/Cache/QueryCache.h [hunk 2]
```diff
@@ -146,11 +147,15 @@ class QueryCache
     {
     public:
         bool hasCacheEntryForKey() const;
-        Pipe && getPipe(); /// must be called only if hasCacheEntryForKey() returns true
+        std::unique_ptr<SourceFromChunks> getSource();
```
[the remark this answers is not in the document]
Comment (rschu1ze) [reply src/Interpreters/Cache/QueryCache.h:152]: Done, thanks.

**Comment by rschu1ze** [comment]
ClickHouse Stateless Tests (release, Wide Parts Enabled):
ClickHouse Stateless Tests (msan) [6/6]:
ClickHouse Stateless Tests (debug) [5/5]:
ClickHouse Stateless Tests (aarch64):
- 02723_zookeeper_name: also occurs in master (missing ORDER BY), will try to fix separately

ClickHouse Stateless Tests (release, Analyzer)
- the many test failures also occur in master (all unrelated)

ClickHouse Integration Tests (release) [2/4]
- test_merge_tree_s3/test.py::test_store_cleanup_disk_s3[node]: #48691

**Comment by rschu1ze** [comment]
> * 02723_zookeeper_name: also occurs in master (missing ORDER BY), will try to fix separately

-- #49594

**Pull request merged**